pkg/build/controller/buildpod/controller.go

```diff
@@ -0,0 +1,401 @@
+		reason := ""
+		if len(build.Status.Reason) > 0 {
+			reason = " (" + string(build.Status.Reason) + ")"
+		}
+		glog.V(4).Infof("Updating build %s/%s status %s -> %s%s", build.Namespace, build.Name, build.Status.Phase, nextStatus, reason)
+		build.Status.Phase = nextStatus
+
+		if buildutil.IsBuildComplete(build) {
+			common.SetBuildCompletionTimeAndDuration(build)
+		}
+		if build.Status.Phase == buildapi.BuildPhaseRunning {
+			now := unversioned.Now()
+			build.Status.StartTimestamp = &now
+		}
+		if err := bc.buildUpdater.Update(build.Namespace, build); err != nil {
+			return fmt.Errorf("failed to update build %s/%s: %v", build.Namespace, build.Name, err)
+		}
+		glog.V(4).Infof("Build %s/%s status was updated to %s", build.Namespace, build.Name, build.Status.Phase)
+
+		if buildutil.IsBuildComplete(build) {
+			common.HandleBuildCompletion(build, bc.runPolicies)
+		}
+	}
+	return nil
+}
+
+// HandleBuildPodDeletion sets the status of a build to error if the build pod has been deleted
+func (bc *BuildPodController) HandleBuildPodDeletion(pod *kapi.Pod) error {
+	glog.V(4).Infof("Handling deletion of build pod %s/%s", pod.Namespace, pod.Name)
+	build, exists, err := bc.getBuildForPod(pod)
+	if err != nil {
+		glog.V(4).Infof("Error getting build for pod %s/%s", pod.Namespace, pod.Name)
+		return err
+	}
+	if !exists || build == nil {
+		glog.V(5).Infof("No build found for deleted pod %s/%s", pod.Namespace, pod.Name)
+		return nil
+	}
+
+	if build.Spec.Strategy.JenkinsPipelineStrategy != nil {
+		glog.V(4).Infof("Build %s/%s is a pipeline build, ignoring", build.Namespace, build.Name)
+		return nil
+	}
+	// If build was cancelled, we'll leave HandleBuild to update the build
+	if build.Status.Cancelled {
+		glog.V(4).Infof("Cancelation for build %s/%s was already triggered, ignoring", build.Namespace, build.Name)
+		return nil
+	}
+
+	if buildutil.IsBuildComplete(build) {
+		glog.V(4).Infof("Pod was deleted but build %s/%s is already completed, so no need to update it.", build.Namespace, build.Name)
+		return nil
+	}
+
+	nextStatus := buildapi.BuildPhaseError
+	if build.Status.Phase != nextStatus {
+		glog.V(4).Infof("Updating build %s/%s status %s -> %s", build.Namespace, build.Name, build.Status.Phase, nextStatus)
+		build.Status.Phase = nextStatus
+		build.Status.Reason = buildapi.StatusReasonBuildPodDeleted
+		build.Status.Message = buildapi.StatusMessageBuildPodDeleted
+		common.SetBuildCompletionTimeAndDuration(build)
+		if err := bc.buildUpdater.Update(build.Namespace, build); err != nil {
+			return fmt.Errorf("Failed to update build %s/%s: %v", build.Namespace, build.Name, err)
+		}
+	}
+	return nil
+}
+
+func (bc *BuildPodController) worker() {
+	for {
+		if quit := bc.work(); quit {
+			return
+		}
+	}
+}
+
+func (bc *BuildPodController) work() bool {
+	key, quit := bc.queue.Get()
+	if quit {
+		return true
+	}
+	defer bc.queue.Done(key)
+
+	pod, err := bc.getPodByKey(key.(string))
+	if err != nil {
+		glog.Error(err.Error())
```

Comment:
You should replace all instances of glog.Errorf with utilruntime.HandleError (it's rate-limited and generally accepted over glog.Errorf)